discovery/file/file.go

```diff
@@ -35,6 +37,65 @@ import (
 
 const fileSDFilepathLabel = model.MetaLabelPrefix + "filepath"
 
+// TimestampCollector is a Custom Collector for Timestamps of the files.
+type TimestampCollector struct {
+	Description *prometheus.Desc
+	discoverers map[*Discovery]struct{}
+	lock        sync.RWMutex
+}
+
+// Describe method sends the description to the channel.
+func (t *TimestampCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- t.Description
+}
+
+// Collect creates constant metrics for each file with last modified time of the file.
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	// new map to dedup dilenames
+	uniqueFiles := make(map[string]float64)
+	t.lock.RLock()
+	discoverers := t.discoverers
```

Comment:
This just copies a reference, you need to hold the lock until after the first for loop.